kuma/core/tests/test_locale_middleware.py

```diff
@@ -1,74 +1,79 @@
+import pytest
 from django.conf import settings
 
-from kuma.core.tests import assert_shared_cache_header
+from . import assert_shared_cache_header
 
-from . import KumaTestCase
 
+# Simple Accept-Language headers, one term
+SIMPLE_ACCEPT_CASES = (
+    ('', 'en-US'),          # No preference gets default en-US
+    ('en', 'en-US'),        # Default en is en-US
+    ('en-US', 'en-US'),     # Exact match for default
+    ('en-us', 'en-US'),     # Case-insensitive match for default
+    ('fr-FR', 'fr'),        # Overly-specified locale gets default
+    ('fr-fr', 'fr'),        # Overly-specified match is case-insensitive
+)
+# Real-world Accept-Language headers include quality value weights
+WEIGHTED_ACCEPT_CASES = (
+    ('en, fr;q=0.5', 'en-US'),          # English without region gets en-US
+    ('en-GB, fr-FR;q=0.5', 'en-US'),    # Any English gets en-US
+    ('en-US, en;q=0.5', 'en-US'),       # Request for en-US gets en-US
+    ('fr, en-US;q=0.5', 'fr'),          # Exact match of non-English language
+    ('fr-FR, de-DE;q=0.5', 'fr'),       # Highest locale-specific match wins
+    ('fr-FR, de;q=0.5', 'de'),          # Highest exact match wins
+    ('ga, fr;q=0.5', 'ga-IE'),          # Generic Gaelic matches ga-IE
+    ('pt, fr;q=0.5', 'pt-PT'),          # Generic Portuguese matches pt-PT
+    ('pt-BR, en-US;q=0.5', 'pt-BR'),    # Portuguese-Brazil matches
+    ('qaz-ZZ, fr-FR;q=0.5', 'fr'),      # Respect partial match on prefix
+    ('qaz-ZZ, qaz;q=0.5', False),       # No matches gets default en-US
+    ('zh-Hant, fr;q=0.5', 'zh-TW'),     # Traditional Chinese matches zh-TW
+)
+PICKER_CASES = SIMPLE_ACCEPT_CASES + WEIGHTED_ACCEPT_CASES + (
+    ('xx', 'en-US'),        # Unknown in Accept-Language gets default
+)
+REDIRECT_CASES = [case for case in SIMPLE_ACCEPT_CASES if case[0] != case[1]]
 
-class TestLocaleMiddleware(KumaTestCase):
-    def test_default_redirect(self):
-        # User wants en-us, we send en-US
-        response = self.client.get('/', follow=True,
-                                   HTTP_ACCEPT_LANGUAGE='en-us')
-        self.assertRedirects(response, '/en-US/', status_code=302)
-        assert_shared_cache_header(response)
 
-        # User wants fr-FR, we send fr
-        response = self.client.get('/', follow=True,
-                                   HTTP_ACCEPT_LANGUAGE='fr-fr')
-        self.assertRedirects(response, '/fr/', status_code=302)
-        assert_shared_cache_header(response)
+@pytest.mark.parametrize('accept_language,locale', PICKER_CASES)
+def test_locale_middleware_picker(accept_language, locale, client, db):
+    '''The LocaleMiddleware picks locale from the Accept-Language header.'''
+    response = client.get('/', HTTP_ACCEPT_LANGUAGE=accept_language)
+    assert response.status_code == 302
+    assert response['Location'] == 'http://testserver/%s/' % locale or 'en_US'
```

Comment:
The `pt` and `zh-Hant` cases make me wonder if the new `LocaleMiddleware` is going to have to handle the `settings.LOCALE_ALIASES` instead of counting on handling them in `LocaleStandardizerMiddleware`?

Reply:
Hmm. The ``or 'en-US'`` was for the case where the client requests unknown locales:

https://github.com/mozilla/kuma/blob/25ddd313d7b0c1b3dd58947fe02cd312da94c110/kuma/core/tests/test_urlresolvers.py#L46-L49

``get_best_language`` returns ``False``, and the redirect URL is the default ``en-US``. That's what I get for trying to be clever...